dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,876 @@
+            parameters (3 eigenvalues, followed by the 3 corresponding
+            eigenvectors) and the free water volume fraction.
+        S0 : float or ndarray
+            The non diffusion-weighted signal in every voxel, or across all
+            voxels. Default: 1
+        """
+        return fwdti_prediction(fwdti_params, self.gtab, S0=S0)
+
+
+class FreeWaterTensorFit(TensorFit):
+    """ Class for fitting the Free Water Tensor Model """
+    def __init__(self, model, model_params, S0):
+        """ Initialize a FreeWaterTensorFit class instance.
+        Since the free water tensor model is an extension of DTI, class
+        instance is defined as subclass of the TensorFit from dti.py
+
+        Parameters
+        ----------
+        model : FreeWaterTensorModel Class instance
+            Class instance containing the free water tensor model for the fit
+        model_params : ndarray (x, y, z, 13) or (n, 13)
+            All parameters estimated from the free water tensor model.
+            Parameters are ordered as follows:
+                1) Three diffusion tensor's eigenvalues
+                2) Three lines of the eigenvector matrix each containing the
+                   first, second and third coordinates of the eigenvector
+                3) The volume fraction of the free water compartment
+        S0 : ndarray (x, y, z) or (n)
+            The free water DTI model estimate of the non diffusion-weighted
+            signal S0
+        """
+        TensorFit.__init__(self, model, model_params)
+        self.S0 = S0
+
+    @property
+    def f(self):
+        """ Returns the free water diffusion volume fraction f """
+        return self.model_params[..., 12]
+
+    def predict(self, gtab, S0=1, step=None):
+        r""" Given a free water tensor model fit, predict the signal on the
+        vertices of a gradient table
+
+        Parameters
+        ----------
+        gtab : a GradientTable class instance
+            The gradient table for this prediction
+
+        S0 : float array
+           The mean non-diffusion weighted signal in each voxel. Default: 1 in
+           all voxels.
+
+        step : int, optional
+            Number of voxels to be processed simultaneously
+        """
+        shape = self.model_params.shape[:-1]
+        size = np.prod(shape)
+        if step is None:
+            step = self.model.kwargs.get('step', size)
+        if step >= size:
+            return fwdti_prediction(self.model_params, gtab, S0=S0)
+        params = np.reshape(self.model_params,
+                            (-1, self.model_params.shape[-1]))
+        predict = np.empty((size, gtab.bvals.shape[0]))
+        for i in range(0, size, step):
+            predict[i:i+step] = fwdti_prediction(params[i:i+step], gtab, S0=S0)
+        return predict.reshape(shape + (gtab.bvals.shape[0],))
+
+
+def _wls_iter(design_matrix, sig, min_diffusivity, min_signal, Diso=3e-3,
```

Comment:
I wonder whether this is the correct name for this function. Is it really doing a WLS iteration?